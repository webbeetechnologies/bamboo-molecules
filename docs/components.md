# Bamboo Molecules

Bamboo Molecules components are built on top of BambooAtoms; thus, at it's core; Bamboo allows for all the components to be composed by default. That means; you can replace an underlying view by a Styled View of your choice; and that is only the beginning.


Bamboo atoms have minimal theming; and Bamboo Molecules are opinionated styles; however, Molecules allow you to customize or entirely change the theme of a specific component. It is possible to use custom components/styles at a module level.


Molecules is heavily inspired by Material and Cupertino components and thus tries to mimic them to provide a native experience as closely as possible on the react native platform.



# References:
- [Comparing iOS and Android UI](https://www.learnui.design/blog/ios-vs-android-app-ui-design-complete-guide.html)
- [Material 3 Figma Design](https://www.figma.com/file/hGlWM0wnP1UN5VtPGiaOJ2/Material-3-Design-Kit-(Community)?node-id=50716%3A11360)
- [Material 3 Design Guidelines](https://m3.material.io/components)

## Molecules HOC:
- [WithAddons](./components/HOC/WithAddons.md)
- [WithAddonAfter](./components/HOC/WithAddons.md#withaddonafter)
- [WithAddonBefore](./components/HOC/WithAddons.md#withaddonbefore)
- [WithElements](./components/HOC/WithElements.md)
- [WithActionIconButton](./components/HOC/WithActionIconButton.md)
- [WithPopover](./components/HOC/WithPopover.md)


## Molecules created using Atoms:
- High Priority
    -   [Icons](./components/Icons.md)
    -   [Input](./components/Input.md)
    -   [Tooltips](./components/Tooltips.md)
    -   Actionsheet
    -   [Flat List](./components/FlatList.md)
    -   Modals/ Dialogs
    -   [Button](./components/Button.md)
    -   [IconButton](./components/IconButton.md)
    -   [Select](./components/Select.md)
    -   [Radio](./components/Radio.md)
    -   [Checkbox](./components/Checkbox.md)
    -   [Date Picker](./components/DatePicker.md)
    -   [Time Picker](./components/TimePicker.md)
    -   Menu and Menu Items
    -   Alert
    -   Cards
    -   Accordion
    -   [Chips](./components/Chips.md)
    -   Drawer


- Medium Priority
    -   [Section List](./components/SectionList.md)
    -   [Range Picker](./components/RangePickers.md)
    -   [Date Time Picker](./components/DateTimePicker.md)
    -   Stack
    -   Divider
    -   Toast
    -   Fab/ Fab Group (Android/Web Only)
    -   Tabs
        -   Pills
        -   Bottom Tabs


- Low Priority
    -   [Color Picker](./components/ColorPicker.md)
    -   [FilePicker](./components/FilePicker.md)
    -   Badges
    -   Search
    -   Forms
    -   Avatar
    -   Graphs




... and more as we identify ..
