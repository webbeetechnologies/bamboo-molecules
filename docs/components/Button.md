# Button
- As a developer, I want to inject the base Button (React-native replacement) component.
- As a user, I want to see feedback from button on hover and focus.
- As a user, I want to know that the button is disabled.
- As a developer, I want to have Icons on left and/or right of the button.

## Implementation Details
- Implements [WithElements HOC](./HOC/WithElements.md)

## References
- [Material 3 Figma Design](https://www.figma.com/file/hGlWM0wnP1UN5VtPGiaOJ2/Material-3-Design-Kit-(Community)?node-id=50716%3A11360)
- [Material 3 Design Guidelines](https://m3.material.io/components/buttons/overview)

## Screenshots
#### Material 3 Button
![Material 3 Buttons](./assets/m3.buttons.png)

# Button Group
- As a developer, I want the ability to group buttons;

## Implementation Details
- Implements [WithAddons HOC](./HOC/WithAddons.md)
- Implements [WithRipple HOC](./HOC/WithRipple.md)

## References
- [Material 3 Segmented Buttons](https://m3.material.io/components/segmented-buttons/overview)

## Screenshots
#### Material 3 Segmented Button
![Material 3 Segmented Buttons](./assets/m3.button-groups.png)