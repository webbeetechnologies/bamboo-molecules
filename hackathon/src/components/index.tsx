export { default as AttributeValueInput } from './AttributeValueInput/AttributeValueInput';
export { default as CustomDrawer } from './CustomDrawer/CustomDrawer';
export { default as MachineCard } from './MachineCard/MachineCard';
export { default as MachineTypeCard } from './MachineTypeCard/MachineTypeCard';
