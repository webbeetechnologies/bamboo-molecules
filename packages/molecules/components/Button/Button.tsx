import { useCallback, ReactNode, memo, useMemo, forwardRef, PropsWithoutRef } from 'react';
import { View, type ViewStyle, type StyleProp, type TextStyle, type ViewProps } from 'react-native';
import setColor from 'color';
import { Text } from '../Text';

import { Icon, type IconType } from '../Icon';
import { Surface, type SurfaceProps } from '../Surface';
import { ButtonSize, ButtonVariant } from './types';
import { TouchableRipple } from '../TouchableRipple';
import { ActivityIndicator } from '../ActivityIndicator';
import { StateLayer } from '../StateLayer';
import { defaultStyles, sizeToIconSizeMap } from './utils';
import { MD3Elevation } from '../../types/theme';
import { resolveStateVariant } from '../../utils';
import { useActionState } from '../../hooks';
import { extractProperties } from '../Surface/utils';

export type Props = Omit<SurfaceProps, 'style'> & {
    /**
     * Mode of the button. You can change the mode to adjust the styling to give it desired emphasis.
     * - `text` - flat button without background or outline, used for the lowest priority actions, especially when presenting multiple options.
     * - `outlined` - button with an outline without background, typically used for important, but not primary action â€“ represents medium emphasis.
     * - `contained` - button with a background color, used for important action, have the most visual impact and high emphasis.
     * - `elevated` - button with a background color and elevation, used when absolutely necessary e.g. button requires visual separation from a patterned background. @supported Available in v5.x with theme version 3
     * - `contained-tonal` - button with a secondary background color, an alternative middle ground between contained and outlined buttons. @supported Available in v5.x with theme version 3
     */
    variant?: ButtonVariant;
    /**
     * @supported Available in v5.x
     * Custom button's background color.
     */
    buttonColor?: string;
    /**
     * Whether to show a loading indicator.
     */
    loading?: boolean;
    /**
     * Icon to display for the `Button`.
     */
    iconType?: IconType;
    iconName?: string;
    iconSize?: number;
    /**
     * Whether the button is disabled. A disabled button is greyed out and `onPress` is not called on touch.
     */
    disabled?: boolean;
    /**
     * Label text of the button.
     */
    children: ReactNode;
    /**
     * Accessibility label for the button. This is read by the screen reader when the user taps the button.
     */
    accessibilityLabel?: string;
    /**
     * Accessibility hint for the button. This is read by the screen reader when the user taps the button.
     */
    accessibilityHint?: string;
    /**
     * Function to execute on press.
     */
    onPress?: () => void;
    /**
     * @supported Available in v5.x
     * Function to execute as soon as the touchable element is pressed and invoked even before onPress.
     */
    onPressIn?: () => void;
    /**
     * @supported Available in v5.x
     * Function to execute as soon as the touch is released even before onPress.
     */
    onPressOut?: () => void;
    /**
     * Function to execute on long press.
     */
    onLongPress?: () => void;
    /**
     * Style of button's inner content.
     * Use this prop to apply custom height and width and to set the icon on the right with `flexDirection: 'row-reverse'`.
     */
    contentStyle?: StyleProp<ViewStyle>;
    style?: StyleProp<TextStyle>;
    /**
     * Style for the button text.
     */
    labelStyle?: TextStyle;
    /**
     * Style for the Icon
     */
    iconContainerStyle?: StyleProp<ViewStyle>;
    /*
     * Size
     * */
    size?: ButtonSize;
    /*
     * Elevation level
     * */
    elevation?: MD3Elevation;
    /**
     * testID to be used on tests.
     */
    testID?: string;
    /**
     * props for the stateLayer
     */
    stateLayerProps?: PropsWithoutRef<ViewProps>;
    textRelatedStyle?: StyleProp<TextStyle>;
};

const elevationMap: Record<string, Record<string, number>> = {
    true: {
        contained: 1,
        'contained-tonal': 1,
        elevated: 2,
    },
    false: {
        elevated: 1,
    },
};

const Button = (
    {
        disabled = false,
        variant = 'text',
        size = 'lg',
        loading,
        iconType,
        iconName,
        buttonColor: customButtonColor,
        children,
        accessibilityLabel,
        accessibilityHint,
        onPress,
        onPressIn,
        onPressOut,
        onLongPress,
        style: styleProp,
        contentStyle,
        labelStyle,
        iconContainerStyle: iconContainerStyleProp,
        testID,
        accessible,
        stateLayerProps = {},
        elevation: elevationProp,
        iconSize: _iconSizeProp,
        textRelatedStyle,
        ...rest
    }: Props,
    ref: any,
) => {
    const { hovered, actionsRef } = useActionState({ ref, actionsToListen: ['hover'] });

    const state = resolveStateVariant({
        disabled,
        hovered,
    });

    defaultStyles.useVariants({
        variant,
        state,
        size,
    });

    // const componentStyles = useComponentStyles(
    //     'Button',
    //     [styleProp, { customButtonColor, customTextColor }],
    //     {
    //         variant,
    //         state: disabled ? 'disabled' : hovered ? 'hovered' : undefined,
    //         size,
    //     },
    // );

    // console.log({ hovered, componentStyles });

    const isVariant = useCallback(
        (variantComponent: ButtonVariant) => {
            return variant === variantComponent;
        },
        [variant],
    );

    const iconSize = _iconSizeProp ?? sizeToIconSizeMap[size] ?? sizeToIconSizeMap.md;
    const elevationLevel = elevationMap[(!!hovered).toString()][variant] ?? 0;

    const {
        customLabelColor,
        customLabelSize,
        rippleColor,
        surfaceStyle,
        textStyle,
        iconStyle,
        viewStyle,
        iconContainerStyle,
        accessibilityState,
        stateLayerStyle,
    } = useMemo(() => {
        const { button, content, icon, iconTextMode, label, labelText, labelTextAddons } =
            defaultStyles;

        // for mobile
        const { borderRadius } = extractProperties(
            [defaultStyles.root, styleProp],
            ['borderRadius'],
        );

        const backgroundColor = customButtonColor && !disabled ? customButtonColor : undefined;

        const _iconStyle = [icon, isVariant('text') && iconTextMode];

        const { color: labelColor, fontSize: labelFontSize } = labelStyle ?? {};

        // TODO - remove this workaround
        let _rippleColor: string | undefined;

        try {
            _rippleColor = setColor(labelColor).alpha(0.12).rgb().string();
        } catch (e) {
            _rippleColor = undefined;
        }

        return {
            customLabelColor: labelColor,
            customLabelSize: labelFontSize,
            rippleColor: _rippleColor,
            surfaceStyle: [
                button,
                backgroundColor ? { backgroundColor } : {},
                defaultStyles.root,
                styleProp,
            ],

            iconStyle: [_iconStyle, textRelatedStyle] as unknown as ViewStyle,
            viewStyle: [
                content,
                { flexGrow: 1 },
                borderRadius ? { borderRadius } : {},
                contentStyle,
            ],
            iconContainerStyle: [iconContainerStyleProp],
            textStyle: [
                isVariant('text') ? (iconName || loading ? labelTextAddons : labelText) : label,
                textRelatedStyle,
                labelStyle,
            ],
            accessibilityState: { disabled },
            stateLayerStyle: [defaultStyles.stateLayer, stateLayerProps?.style],
        };
        // eslint-disable-next-line
    }, [
        state,
        variant,
        size,
        contentStyle,
        customButtonColor,
        disabled,
        iconContainerStyleProp,
        iconName,
        isVariant,
        labelStyle,
        loading,
        stateLayerProps?.style,
        styleProp,
    ]);

    const elevation = useMemo(
        () => (elevationProp === undefined ? elevationLevel ?? 0 : elevationProp),
        [elevationLevel, elevationProp],
    );

    return (
        <Surface
            {...rest}
            style={surfaceStyle}
            elevation={
                (disabled
                    ? 0
                    : hovered
                    ? (elevationProp || 0) + elevationLevel
                    : elevation) as MD3Elevation
            }>
            <TouchableRipple
                borderless
                onPress={onPress}
                onLongPress={onLongPress}
                onPressIn={onPressIn}
                onPressOut={onPressOut}
                accessibilityLabel={accessibilityLabel}
                accessibilityHint={accessibilityHint}
                accessibilityRole="button"
                accessibilityState={accessibilityState}
                accessible={accessible}
                disabled={disabled}
                rippleColor={rippleColor}
                style={viewStyle}
                ref={actionsRef}
                testID={testID}>
                <>
                    {iconName && loading !== true ? (
                        <View style={iconContainerStyle}>
                            <Icon
                                type={iconType}
                                name={iconName}
                                size={customLabelSize ?? iconSize}
                                color={
                                    typeof customLabelColor === 'string'
                                        ? customLabelColor
                                        : undefined
                                }
                                style={iconStyle}
                            />
                        </View>
                    ) : null}
                    {loading ? (
                        <ActivityIndicator
                            size={customLabelSize ?? iconSize}
                            color={
                                (typeof customLabelColor === 'string'
                                    ? customLabelColor
                                    : undefined) as string
                            }
                            style={iconStyle}
                        />
                    ) : null}
                    <Text selectable={false} numberOfLines={1} style={textStyle}>
                        {children}
                    </Text>

                    <StateLayer
                        testID={testID ? `${testID}-stateLayer` : ''}
                        {...stateLayerProps}
                        style={stateLayerStyle}
                    />
                </>
            </TouchableRipple>
        </Surface>
    );
};

export default memo(forwardRef(Button));
