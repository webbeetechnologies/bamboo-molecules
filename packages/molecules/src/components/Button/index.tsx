export { default as Button, Props as ButtonProps } from './Button';
export { defaultStyles as buttonStyles } from './utils';
export * from './types';
