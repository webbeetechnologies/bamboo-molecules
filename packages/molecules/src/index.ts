export { withRipple, withActionState, CallbackActionState, typedMemo } from './hocs';

export {
    useTheme,
    useColorMode,
    useCurrentTheme,
    useToggle,
    useMediaQuery,
    useControlledValue,
    useFilePicker,
    useLatest,
    useSearchable,
    useSearchInputProps,
    UseSearchableProps,
    usePrevious,
    useMergedRefs,
    useSubcomponents,
    useBackHandler,
    useKeyboardDismissable,
    // react-native-web-hooks
    useActive,
    useFocus,
    useHover,
    useREM,
    useLayout,
    useDimensions,
    useScaledSize,
    useMaskedInputProps,
    UseMaskedInputProps,
    useQueryFilter,
    UseQueryFilterProps,
    useHandleNumberFormat,
    UseHandleNumberFormatProps,
    NumberMaskConfig,
    useActionState,
    useContrastColor,
} from './hooks';

export {
    MD3LightTheme,
    MD3DarkTheme,
    tokens,
    generateLightThemeColors,
    generateDarkThemeColors,
    lightenBy,
    darkenBy,
} from './styles';

export {
    normalizeSpacings,
    DocumentPicker,
    SpacingType,
    SpacingKey,
    DocumentPickerOptions,
    DocumentResult,
    documentTypes,
    BackgroundContext,
    formatNumberWithMask,
    normalizeToNumberString,
    isMac,
    getOS,
    color,
    resolveContrastColor,
} from './utils';

export * from './types';
