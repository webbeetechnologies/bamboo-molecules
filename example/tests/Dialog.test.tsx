import { Dialog } from '../../src/components';
import { Example as Button } from '../stories/components/Button/Button';
import { renderWithWrapper } from '../testHelpers';
import { Text } from 'react-native';

describe('Dialog', () => {
    it('renders with content', () => {
        const tree = renderWithWrapper(
            <Dialog isOpen>
                <Dialog.Content>
                    <Text>This is content</Text>
                </Dialog.Content>
            </Dialog>,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('renders with title, content and 2 action buttons', () => {
        const tree = renderWithWrapper(
            <Dialog isOpen testID={'dialogTest'}>
                <Dialog.Title>Dialog Title</Dialog.Title>
                <Dialog.Content>
                    <Text>This is content</Text>
                </Dialog.Content>
                <Dialog.Actions>
                    <Button>Cancel</Button>
                    <Button>Save</Button>
                </Dialog.Actions>
            </Dialog>,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
