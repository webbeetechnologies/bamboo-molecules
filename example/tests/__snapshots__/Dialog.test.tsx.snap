// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog renders with content 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  presentationStyle="overFullScreen"
  statusBarTranslucent={true}
  supportedOrientations={
    Array [
      "portrait",
      "portrait-upside-down",
      "landscape",
      "landscape-left",
      "landscape-right",
    ]
  }
  transparent={true}
  visible={true}
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 1)",
        "bottom": 0,
        "flex": 1,
        "left": 0,
        "opacity": 0.3,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <View
    pointerEvents="box-none"
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "flex": 1,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "alignSelf": undefined,
          "bottom": undefined,
          "left": undefined,
          "position": undefined,
          "right": undefined,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.15,
          "shadowRadius": 6,
          "top": undefined,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.3,
            "shadowRadius": 2,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "rgba(255, 251, 254, 1)",
              "borderRadius": 28,
              "justifyContent": "flex-start",
              "maxHeight": 600,
              "maxWidth": 560,
              "minWidth": 280,
              "overflow": "hidden",
              "width": 750,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
                "marginTop": 24,
                "paddingBottom": 24,
                "paddingHorizontal": 24,
              }
            }
          >
            <Text>
              This is content
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;

exports[`Dialog renders with title, content and 2 action buttons 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  presentationStyle="overFullScreen"
  statusBarTranslucent={true}
  supportedOrientations={
    Array [
      "portrait",
      "portrait-upside-down",
      "landscape",
      "landscape-left",
      "landscape-right",
    ]
  }
  testID="dialogTest"
  transparent={true}
  visible={true}
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 1)",
        "bottom": 0,
        "flex": 1,
        "left": 0,
        "opacity": 0.3,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <View
    pointerEvents="box-none"
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "flex": 1,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "alignSelf": undefined,
          "bottom": undefined,
          "left": undefined,
          "position": undefined,
          "right": undefined,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.15,
          "shadowRadius": 6,
          "top": undefined,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.3,
            "shadowRadius": 2,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "rgba(255, 251, 254, 1)",
              "borderRadius": 28,
              "justifyContent": "flex-start",
              "maxHeight": 600,
              "maxWidth": 560,
              "minWidth": 280,
              "overflow": "hidden",
              "width": 750,
            }
          }
        >
          <Text
            accessibilityRole="header"
            style={
              Object {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontSize": 24,
                "fontWeight": "400",
                "letterSpacing": 0,
                "lineHeight": 32,
                "marginHorizontal": 24,
                "marginTop": 24,
                "marginVertical": 12,
              }
            }
          >
            Dialog Title
          </Text>
          <View
            style={
              Object {
                "flex": 1,
                "paddingBottom": 24,
                "paddingHorizontal": 24,
              }
            }
          >
            <Text>
              This is content
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "flex-end",
                "paddingBottom": 24,
                "paddingHorizontal": 24,
              }
            }
          >
            <View>
              <View
                collapsable={false}
                style={
                  Object {
                    "alignSelf": undefined,
                    "bottom": undefined,
                    "left": undefined,
                    "position": undefined,
                    "right": undefined,
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 0,
                      "width": 0,
                    },
                    "shadowOpacity": 0,
                    "shadowRadius": 0,
                    "top": undefined,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "shadowColor": "#000",
                      "shadowOffset": Object {
                        "height": 0,
                        "width": 0,
                      },
                      "shadowOpacity": 0,
                      "shadowRadius": 0,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    focused={false}
                    pressed={false}
                    style={
                      Object {
                        "backgroundColor": "transparent",
                        "borderRadius": 750,
                        "borderStyle": "solid",
                        "flex": 1,
                        "marginLeft": 0,
                        "minWidth": 64,
                      }
                    }
                  >
                    <View
                      accessibilityRole="button"
                      accessibilityState={
                        Object {
                          "disabled": true,
                        }
                      }
                      accessible={true}
                      focusable={false}
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        Array [
                          Object {
                            "overflow": "hidden",
                          },
                          Object {
                            "borderRadius": 750,
                            "flex": 1,
                          },
                        ]
                      }
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "alignItems": "center",
                              "flex": 1,
                              "flexDirection": "row",
                              "justifyContent": "center",
                            },
                            undefined,
                          ]
                        }
                      >
                        <Text
                          numberOfLines={1}
                          selectable={false}
                          style={
                            Object {
                              "alignItems": "center",
                              "color": "",
                              "display": "flex",
                              "fontFamily": "System",
                              "fontSize": 16,
                              "fontWeight": "500",
                              "letterSpacing": 0.1,
                              "lineHeight": 20,
                              "marginHorizontal": 12,
                              "marginVertical": 10,
                              "textAlign": "center",
                            }
                          }
                        >
                          Cancel
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
            <View>
              <View
                collapsable={false}
                style={
                  Object {
                    "alignSelf": undefined,
                    "bottom": undefined,
                    "left": undefined,
                    "position": undefined,
                    "right": undefined,
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 0,
                      "width": 0,
                    },
                    "shadowOpacity": 0,
                    "shadowRadius": 0,
                    "top": undefined,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "shadowColor": "#000",
                      "shadowOffset": Object {
                        "height": 0,
                        "width": 0,
                      },
                      "shadowOpacity": 0,
                      "shadowRadius": 0,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    focused={false}
                    pressed={false}
                    style={
                      Object {
                        "backgroundColor": "transparent",
                        "borderRadius": 750,
                        "borderStyle": "solid",
                        "flex": 1,
                        "marginLeft": 8,
                        "minWidth": 64,
                      }
                    }
                  >
                    <View
                      accessibilityRole="button"
                      accessibilityState={
                        Object {
                          "disabled": true,
                        }
                      }
                      accessible={true}
                      focusable={false}
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        Array [
                          Object {
                            "overflow": "hidden",
                          },
                          Object {
                            "borderRadius": 750,
                            "flex": 1,
                          },
                        ]
                      }
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "alignItems": "center",
                              "flex": 1,
                              "flexDirection": "row",
                              "justifyContent": "center",
                            },
                            undefined,
                          ]
                        }
                      >
                        <Text
                          numberOfLines={1}
                          selectable={false}
                          style={
                            Object {
                              "alignItems": "center",
                              "color": "",
                              "display": "flex",
                              "fontFamily": "System",
                              "fontSize": 16,
                              "fontWeight": "500",
                              "letterSpacing": 0.1,
                              "lineHeight": 20,
                              "marginHorizontal": 12,
                              "marginVertical": 10,
                              "textAlign": "center",
                            }
                          }
                        >
                          Save
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
